README.md
# openapi-mock-middleware

## Installation

Generating mock data using open API file

## Usage

npm i openapi-mock-middleware

```js
const express = require('express');
const createMockMiddleware = require('openapi-mock-middleware');

const app = express();

app.use(
  '/api',
  createMockMiddleware({
    request: {
      spec: ['./example'], // apiSpec。支持file, dir,  array。
      ignore: [/\.js$/], // 扫描目录时忽略的文件 string, RegExp or array
      ext: ['.yml', '.json'], // API文件后缀格式：string or array
      isValidate: true, // 是否开启校验: 开启之后将校验请求的 header、path、query、body。
    },
    response: {
      locale: 'zh', // json-schema-faker locale, default to 'en'
      options: {
        // json-schema-faker options
        /**
         * alwaysFakeOptionals-当启用时，它会设置optionalsProbability: 1.0和 fixedProbabilities: true（默认值：false）
         * optionalsProbability— 一个值 from 0.0to1.0以一致的方式生成值，例如0.5将生成 from 0%to 50%of 的值。使用数组这意味着项目，在对象上他们的属性等。（默认值：false）
         * fixedProbabilities-如果启用，那么optionalsProbability: 0.5总是会生成值的一半（默认值：false）
         */
        alwaysFakeOptionals: true,
        useExamplesValue: true, //如果启用，它会返回一个随机值从examples它们是否存在
        // useDefaultValue-如果启用，它将返回default值，如果存在（默认：false）
      },
      callback: (jsf, faker) => {
        // function where you can extend json-schema-faker
        // ...
      },
    },
  })
);

app.listen(8090, () =>
  console.log('Server listening on http://localhost:8090')
);
```

参考：【[openapi-mock-express-middleware](https://github.com/aleksandryackovlev/openapi-mock-express-middleware)】


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.11](https://github.com/ma-x-x/openapi-mock-middleware/compare/v1.0.10...v1.0.11) (2022-06-02)

### [1.0.10](https://github.com/ma-x-x/openapi-mock-middleware/compare/v1.0.9...v1.0.10) (2022-05-06)

### [1.0.8](https://github.com/ma-x-x/openapi-mock-middleware/compare/v1.0.9...v1.0.8) (2022-05-06)

### [1.0.9](https://github.com/ma-x-x/openapi-mock-middleware/compare/v1.0.8...v1.0.9) (2022-04-01)

### [1.0.8](https://github.com/ma-x-x/openapi-mock-middleware/compare/v1.0.7...v1.0.8) (2022-04-01)

### [1.0.7](https://github.com/ma-x-x/openapi-mock-middleware/compare/v1.0.6...v1.0.7) (2022-03-31)


src/test.ts
const express = require('express');
const { createMockMiddleware } = require('./index');
const path = require('path');
const fs = require('fs');
const { getSync } = require('./utils');
// import express from 'express';
// import { createMockMiddleware } from './dist/index';

// console.log(getSync('../src'));

const app = express();

app.use(
  '/api',
  createMockMiddleware({
    request: {
      spec: ['./example'], // apiSpec。支持file, dir,  array。
      ignore: [/\.js$/], // 扫描目录时忽略的文件 string, RegExp or array
      ext: ['yml', '.json'], // API文件后缀格式：string or array
      isValidate: false, // 是否开启校验: 开启之后将校验请求的 header、path、query、body。
    },
    response: {
      locale: 'zh_CN', // json-schema-faker locale, default to 'en'
      options: {
        // json-schema-faker options
        /**
         * alwaysFakeOptionals-当启用时，它会设置optionalsProbability: 1.0和 fixedProbabilities: true（默认值：false）
         * optionalsProbability— 一个值 from 0.0to1.0以一致的方式生成值，例如0.5将生成 from 0%to 50%of 的值。使用数组这意味着项目，在对象上他们的属性等。（默认值：false）
         * fixedProbabilities-如果启用，那么optionalsProbability: 0.5总是会生成值的一半（默认值：false）
         */

        alwaysFakeOptionals: true,
        useExamplesValue: true, //如果启用，它会返回一个随机值从examples它们是否存在
        // useDefaultValue-如果启用，它将返回default值，如果存在（默认：false）
      },
      callback: (jsf: any, faker: any) => {
        // function where you can extend json-schema-faker
        // ...
      },
    },
  })
);

app.listen(8090, () =>
  console.log('Server listening on http://localhost:8090')
);


src/index.spec.ts
import path from 'path';

import { Router } from 'express';

import { createMockMiddleware } from './index';

describe('createMockMiddleware', () => {
  it('should return an instance of the express router', async () => {
    const middleware = createMockMiddleware({
      request: {
        spec: path.resolve(__dirname, '../test/fixtures/petstore.yaml'),
      },
    });

    expect(Object.getPrototypeOf(middleware)).toBe(Router);
  });

  it('should throw an error if the given file does not exist', () => {
    try {
      createMockMiddleware({
        request: {
          spec: path.resolve(
            __dirname,
            '../test/fixtures/petstore_not_exist.yaml'
          ),
        },
      });
      throw new Error('exit');
    } catch (error) {
      expect(error).toBeInstanceOf(Error);
      expect(error).toHaveProperty(
        'message',
        `OpenAPI spec not found at location: ${path.resolve(
          __dirname,
          '../test/fixtures/petstore_not_exist.yaml'
        )}`
      );
    }
  });
});


src/index.ts
/* eslint-disable no-console */
import fs from 'fs';

import express from 'express';
import type { OpenAPIV3 } from 'openapi-types';

import createRouter from './router';
import { createOperations } from './operations';
import {
  isAuthenticated,
  validateHeaders,
  validatePath,
  validateQuery,
  validateBody,
} from './middleware';
import { requestType, responseType } from './utils';

export interface MiddlewareOptions {
  request: requestType;
  response?: responseType;
}

/** */

export const createMockMiddleware = ({request,response}: MiddlewareOptions): express.Router => {
  if (typeof request.spec === 'string' && !fs.existsSync(request.spec)) {
    throw new Error(`api spec file not exit: ${request.spec}`);
  } else if (request.spec === undefined) {
    throw new Error(`api spec not exit`);
  }

  const router = createRouter();
  const operations = createOperations({
    request,
    response,
  });

  router.use('/{0,}', async (req, res, next) => {
    res.locals.operation = await operations.match(req);
    next();
  });

  /** 开启校验 */
  if (request.isValidate) {
    router.use(isAuthenticated);
    router.use(validateHeaders);
    router.use(validatePath);
    router.use(validateQuery);
    router.use(validateBody);
  }

  router.use((req, res, next) => {
    return res.locals.operation
      ? res.locals.operation.generateResponse(req, res, response?.withResponse)
      : next();
  });

  router.use((req, res) => {
    res.status(404).send({ message: 'Not found' });
  });

  router.use(
    (err: Error, req: express.Request, res: express.Response): void => {
      res.status(500).send({ message: 'Something broke!' });
    }
  );

  return router;
};


example/utils.js
var nodePath = require('path'),
  sep = nodePath.sep,
  fs = require('fs');

/**
 * Recurses through the supplied paths, returning an array of file paths found
 * in each. Paths may be a string, or an array of strings. The paths may be
 * strings pointing to files or directories. Accepts an options object with
 * the following keys: ignore and ext. opts.ignore takes either a single
 * string pattern or array of patterns to ignore. opts.ext accepts either
 * a string, or array of strings corresponding to filename extensions to
 * retrieve.
 *
 * @param {string|string[]} paths The paths to traverse
 * @param {object}          opts  Options object with ignore and ext keys
 */
exports.getSync = function (paths, opts) {
  var results, ignorePatterns, i;

  paths = paths || [];
  opts = opts || {};
  results = [];

  if (!(paths instanceof Array)) {
    paths = [paths];
  }

  ignorePatterns = [];
  if (opts.ignore) {
    if (!(opts.ignore instanceof Array)) {
      opts.ignore = [opts.ignore];
    }

    for (i = 0; i < opts.ignore.length; i++) {
      ignorePatterns.push(new RegExp(opts.ignore[i]));
    }
  }

  if (!opts.ext) {
    opts.ext = [];
  } else if (!(opts.ext instanceof Array)) {
    opts.ext = [opts.ext];
  }

  paths.forEach(function (path) {
    if (!fs.existsSync(path)) {
      throw new Error('No such file or directory: ' + path);
    } else if (fs.statSync(path).isFile()) {
      return results.push(path);
    }

    walkSync(path, function (dirPath, dirs, files) {
      files.forEach(function (file) {
        var filePath;
        var ext = nodePath.extname(file);

        if (opts.ext.length && opts.ext.indexOf(ext) === -1) {
          return;
        }

        if (dirPath.slice(-1) !== sep) {
          dirPath += sep;
        }

        if (dirPath.indexOf(sep) !== 0 && dirPath.indexOf('.') !== 0) {
          dirPath = './' + dirPath;
        }

        filePath = dirPath + file;

        for (var i = 0; i < ignorePatterns.length; i++) {
          if (ignorePatterns[i].test(filePath)) return;
        }

        results.push(filePath);
      });
    });
  });

  return results;
};

function walkSync(dir, fn) {
  if (!fs.statSync(dir).isDirectory()) {
    throw new Error(dir + ' is not a directory');
  }

  var dirs = [];
  var files = [];

  fs.readdirSync(dir).forEach(function (name) {
    var stat = fs.lstatSync(nodePath.join(dir, name));
    // Don't follow symbolic links
    if (stat.isSymbolicLink()) {
      return;
    } else if (stat.isDirectory()) {
      dirs.push(name);
    } else {
      files.push(name);
    }
  });

  fn(dir, dirs, files);

  dirs.forEach(function (subdir) {
    walkSync(nodePath.join(dir, subdir), fn);
  });
}


src/utils/filePaths.ts
import nodePath from 'path';
import fs from 'fs';

const sep = nodePath.sep;

type optsType =
  | {
      ignore?: string | string[] | RegExp | RegExp[];
      ext?: string | string[];
    }
  | undefined;

export const getSync = (paths: string | string[] = [], opts: optsType = {}) => {
  const results: string[] = [];
  const ignorePatterns: RegExp[] = [];
  let i: number;

  if (!(paths instanceof Array)) {
    paths = [paths];
  }

  if (opts.ignore) {
    if (!(opts.ignore instanceof Array)) {
      opts.ignore = [opts.ignore as string];
    }

    for (i = 0; i < opts.ignore.length; i++) {
      ignorePatterns.push(new RegExp(opts.ignore[i]));
    }
  }

  if (!opts.ext) {
    opts.ext = [];
  } else if (!(opts.ext instanceof Array)) {
    opts.ext = [opts.ext];
  }

  paths.forEach(function (path) {
    if (!fs.existsSync(path)) {
      throw new Error('No such file or directory: ' + path);
    } else if (fs.statSync(path).isFile()) {
      return results.push(path);
    }

    walkSync(path, function (dirPath, dirs, files) {
      files.forEach(function (file) {
        let filePath;
        let ext = nodePath.extname(file);

        // @ts-ignore
        if (opts.ext.length && opts.ext.indexOf(ext) === -1) {
          return;
        }

        if (dirPath.slice(-1) !== sep) {
          dirPath += sep;
        }

        if (dirPath.indexOf(sep) !== 0 && dirPath.indexOf('.') !== 0) {
          dirPath = './' + dirPath;
        }

        filePath = dirPath + file;

        for (let i = 0; i < ignorePatterns.length; i++) {
          if (ignorePatterns[i].test(filePath)) return;
        }
        results.push(filePath);
      });
    });
  });

  return results;
};

function walkSync(
  dir: string,
  fn: (dir: string, dirs: string[], files: string[]) => void
) {
  if (!fs.statSync(dir).isDirectory()) {
    throw new Error(dir + ' is not a directory');
  }

  const dirs: string[] = [];
  const files: string[] = [];

  fs.readdirSync(dir).forEach(function (name) {
    const stat = fs.lstatSync(nodePath.join(dir, name));
    // Don't follow symbolic links
    if (stat.isSymbolicLink()) {
      return;
    } else if (stat.isDirectory()) {
      dirs.push(name);
    } else {
      files.push(name);
    }
  });

  fn(dir, dirs, files);

  dirs.forEach(function (subDir) {
    walkSync(nodePath.join(dir, subDir), fn);
  });
}


src/utils/generator.ts
import jsf, { JSF, JSFOptions } from 'json-schema-faker';
import { faker } from '@faker-js/faker';
import type { OpenAPIV3 } from 'openapi-types';

export { JSONSchema, JSFOptions, JSF } from 'json-schema-faker';

export type JSFCallback = (jsfInstance: JSF, fakerObject: typeof faker) => void;
export type requestType = {
  spec: string | string[] | OpenAPIV3.Document; // api文件file, dir,  array or OpenAPIV3.Document object
  ignore?: string | string[] | RegExp | RegExp[]; // 扫描目录时忽略的文件 string, RegExp or array
  ext?: string | string[]; // API文件后缀格式：string or array
  isValidate?: boolean; // 是否开启校验: 开启之后将校验请求的 header、path、query、body。
};
export type responseType = Partial<{
  locale: string;
  options: Partial<JSFOptions>;
  callback: JSFCallback;
  withResponse?: <T>(data: T) => any;
}>;

const defaultOptions = {
  optionalsProbability: 0.5,
};

export const handleExamples = (value: any): any => {
  if (
    typeof value === 'object' &&
    value !== null &&
    Object.keys(value).length
  ) {
    return value[Object.keys(value)[0]].value;
  }

  return '';
};

export const createGenerator: (response?: responseType) => JSF = (response) => {
  const locale = response?.locale || 'zh_CN';
  const options = response?.options || defaultOptions;
  const callback = response?.callback || <JSFCallback>(() => {});
  jsf.option({
    ...defaultOptions,
    ...options,
  });

  jsf.extend('faker', () => {
    const faker = require('@faker-js/faker/locale/zh_CN');
    return faker;
  });

  jsf.define('example', (value) => {
    return value;
  });

  jsf.define('examples', handleExamples);

  return jsf;
};


src/operations/operation.ts
import { OpenAPIV3 } from 'openapi-types';
import express from 'express';
import { has, get, set, findKey } from 'lodash';

import { pathToRegexp } from 'path-to-regexp';

import { JSF, JSONSchema } from '../utils';

export interface ParamsSchemas {
	header: JSONSchema;
	query: JSONSchema;
	path: JSONSchema;
}

function isReferenceObject(response: unknown): response is OpenAPIV3.ReferenceObject {
	return typeof response === 'object' && response !== null && '$ref' in response;
}

export class Operation {
	method: string;

	pathRegexp: RegExp;

	operation: OpenAPIV3.OperationObject;

	pathPattern: string;

	generator: JSF;

	securitySchemes: { [key: string]: OpenAPIV3.SecuritySchemeObject } | null;

	constructor({
		method,
		path,
		operation,
		securitySchemes,
		generator
	}: {
		path: string;
		method: string;
		operation: OpenAPIV3.OperationObject;
		generator: JSF;
		securitySchemes?: { [key: string]: OpenAPIV3.SecuritySchemeObject };
	}) {
		this.pathPattern = path.replace(/\{([^/}]+)\}/g, (p1: string, p2: string): string => `:${p2}`);

		this.method = method.toUpperCase();
		this.operation = operation;
		this.securitySchemes = securitySchemes || null;
		this.generator = generator;

		this.pathRegexp = pathToRegexp(this.pathPattern);
	}

	getResponseStatus(): number {
		const responses = get(this.operation, 'responses');

		if (!responses) {
			return 200;
		}

		const status: string | undefined = findKey(responses, (content, code) => {
			const statusCode = parseInt(code, 10);

			if (Number.isNaN(statusCode)) {
				return false;
			}

			return statusCode >= 200 && statusCode < 299;
		});

		return status ? parseInt(status, 10) : 200;
	}

	getResponseSchema(responseStatus = 200): JSONSchema | null {
		if (
			has(this.operation, [ 'responses', responseStatus, 'content', 'application/json', 'schema' ]) ||
			has(this.operation, [ 'responses', '200', 'schema' ])
		) {
			const { schema, example, examples } =
				get(this.operation, [ 'responses', responseStatus, 'content', 'application/json' ]) ||
				get(this.operation, [ 'responses', '200' ]);

			if (schema && !isReferenceObject(schema)) {
				const resultSchema: JSONSchema = schema as JSONSchema;

				if (example) {
					resultSchema.example = example;
				}

				if (examples) {
					resultSchema.examples = examples;
				}

				return resultSchema;
			}

			return null;
		}

		return null;
	}

	getSecurityRequirements(): OpenAPIV3.SecurityRequirementObject[] {
		const requirements: OpenAPIV3.SecurityRequirementObject[] = this.operation.security || [];

		return requirements;
	}

	getParamsSchemas(): ParamsSchemas {
		const schemas: ParamsSchemas = {
			header: {
				type: 'object',
				required: []
			},
			query: {
				type: 'object',
				additionalProperties: false,
				required: []
			},
			path: {
				type: 'object',
				additionalProperties: false,
				required: []
			}
		};

		const parameters = get(this.operation, [ 'parameters' ]);

		if (parameters) {
			parameters.forEach((parameter) => {
				if (
					parameter &&
					!isReferenceObject(parameter) &&
					(parameter.in === 'header' || parameter.in === 'query' || parameter.in === 'path') &&
					schemas[parameter.in]
				) {
					const prevRequired: string[] = schemas[parameter.in].required || [];

					set(
						schemas,
						[
							parameter.in,
							'properties',
							parameter.in === 'header' ? parameter.name.toLowerCase() : parameter.name
						],
						parameter.schema
					);

					if (parameter.required) {
						set(
							schemas,
							[ parameter.in, 'required' ],
							[
								...prevRequired,
								parameter.in === 'header' ? parameter.name.toLowerCase() : parameter.name
							]
						);
					}
				}
			});
		}

		return schemas;
	}

	getBodySchema(contentType: string): JSONSchema | null {
		return get(this.operation, [ 'requestBody', 'content', contentType, 'schema' ]);
	}

	generateResponse(
		req: express.Request,
		res: express.Response,
		withResponse?: <T>(data: T) => any
	): express.Response {
		const responseStatus = this.getResponseStatus();
		const responseSchema = this.getResponseSchema(responseStatus);
		const responseData = responseSchema ? this.generator.generate(responseSchema) : {};
		return res.status(responseStatus).json(withResponse ? withResponse(responseData) : responseData);
	}
}

export const createOperation = ({
	method,
	path,
	operation,
	generator,
	securitySchemes
}: {
	path: string;
	method: string;
	operation: OpenAPIV3.OperationObject;
	generator: JSF;
	securitySchemes?: { [key: string]: OpenAPIV3.SecuritySchemeObject };
}): Operation =>
	new Operation({
		method,
		path,
		operation,
		generator,
		securitySchemes
	});


src/operations/operations.ts
import path from 'path';
import { OpenAPIV3 } from 'openapi-types';
import chokidar from 'chokidar';
import { Request } from 'express';
import SwaggerParser from '@apidevtools/swagger-parser';
import { get, toPairs } from 'lodash';
import {
  createGenerator,
  JSFOptions,
  JSF,
  JSFCallback,
  requestType,
  responseType,
} from '../utils';
import { Operation, createOperation } from './operation';
import { getSync } from '../utils/filePaths';

export class Operations {
  operations: Operation[] | null = null;

  spec: string | string[] | OpenAPIV3.Document;
  ignore?: string | string[] | RegExp | RegExp[];
  ext?: string | string[];
  generator: JSF;

  constructor({
    request,
    response,
  }: {
    request: requestType;
    response?: responseType;
  }) {
    const { spec, ignore, ext } = request;
    this.spec = spec;
    this.ignore = ignore;
    this.ext = ext;
    this.watch();
    this.generator = createGenerator(response);
  }

  async parseFileToApiRoutersArr(): Promise<OpenAPIV3.Document<{}>[]> {
    let routers = [];

    if (typeof this.spec === 'string') {
      console.log('api files：', this.spec);
      const api = (await SwaggerParser.dereference(
        this.spec
      )) as OpenAPIV3.Document<{}>;
      routers.push(api);
    } else if (Array.isArray(this.spec)) {
      let filesPath = getSync(JSON.parse(JSON.stringify(this.spec)), {
        ignore: this.ignore,
        ext: this.ext,
      });
      if (filesPath.length === 0) {
        throw new Error(`api spec file not exit: ${this.spec}`);
      }
      console.log('api files：', filesPath);
      for (let i = 0; i < filesPath.length; i++) {
        const item = filesPath[i];
        const apiInfo = (await SwaggerParser.dereference(
          item
        )) as OpenAPIV3.Document<{}>;
        routers.push(apiInfo);
      }
    }

    return routers;
  }

  reset(): void {
    this.operations = null;
  }

  watch(): void {
    if (typeof this.spec !== 'string' && !Array.isArray(this.spec)) return;

    // 一个字符串或者是一个数组，描述监听的文件或者文件夹的路径
    const watcher = chokidar.watch(this.spec, {
      ignored: this.ignore, // ignore dotfiles
    });
    watcher.on('all', () => {
      console.log('hot mock................');
      this.reset();
    });
  }

  async compile(): Promise<void> {
    let operations: Operation[] = [];
    let apiRoutersArr = await this.parseFileToApiRoutersArr();
    for (let i = 0; i < apiRoutersArr.length; i++) {
      let item = apiRoutersArr[i];
      let curApiRouters = toPairs(item.paths as OpenAPIV3.PathsObject).reduce(
        (result: Operation[], [pathName, pathOperations]) => [
          ...result,
          ...this.compileFromPath(
            pathName,
            pathOperations as OpenAPIV3.Document,
            get(item, 'components.securitySchemes')
          ),
        ],
        []
      );
      operations = [...operations, ...curApiRouters];
    }

    this.operations = operations;
  }

  /* eslint-disable class-methods-use-this */
  compileFromPath(
    pathName: string,
    pathOperations: OpenAPIV3.PathItemObject,
    securitySchemes?: { [key: string]: OpenAPIV3.SecuritySchemeObject }
  ): Operation[] {
    return toPairs(pathOperations).map(([method, operation]) =>
      createOperation({
        method,
        path: pathName,
        operation: operation as OpenAPIV3.OperationObject,
        securitySchemes,
        generator: this.generator,
      })
    );
  }
  /* eslint-enable class-methods-use-this */

  async match(req: Request): Promise<Operation | null> {
    if (!this.operations) {
      await this.compile();
    }

    const { method, path: pathname } = req;

    return (
      (this.operations &&
        this.operations.find(
          ({ method: operationMethod, pathRegexp }) =>
            pathRegexp.exec(pathname) &&
            method.toUpperCase() === operationMethod.toUpperCase()
        )) ||
      null
    );
  }
}

export const createOperations = ({
  request,
  response,
}: {
  request: requestType;
  response?: responseType;
}): Operations => new Operations({ request, response });


example/example.js
const express = require('express');
const { createMockMiddleware } = require('../dist/index');
const path = require('path');
const fs = require('fs');
const { getSync } = require('./utils');
// import express from 'express';
// import { createMockMiddleware } from './dist/index';

// console.log(getSync('../src'));

const app = express();

app.use(
  '/api',
  createMockMiddleware({
    request: {
      spec: ['./example'], // apiSpec。支持file, dir,  array。
      ignore: [/\.js$/], // 扫描目录时忽略的文件 string, RegExp or array
      ext: ['.yml', '.json'], // API文件后缀格式：string or array
      isValidate: false, // 是否开启校验: 开启之后将校验请求的 header、path、query、body。
    },
    response: {
      locale: 'zh', // json-schema-faker locale, default to 'en'
      options: {
        // json-schema-faker options
        /**
         * alwaysFakeOptionals-当启用时，它会设置optionalsProbability: 1.0和 fixedProbabilities: true（默认值：false）
         * optionalsProbability— 一个值 from 0.0to1.0以一致的方式生成值，例如0.5将生成 from 0%to 50%of 的值。使用数组这意味着项目，在对象上他们的属性等。（默认值：false）
         * fixedProbabilities-如果启用，那么optionalsProbability: 0.5总是会生成值的一半（默认值：false）
         */
        alwaysFakeOptionals: false,
        useExamplesValue: false, //如果启用，它会返回一个随机值从examples它们是否存在
        // useDefaultValue-如果启用，它将返回default值，如果存在（默认：false）
      },
      callback: (jsf, faker) => {
        // function where you can extend json-schema-faker
        // ...
      },
      withResponse: (data) => ({code:'9999',message:'测试一下',data}), // 可以对返回的数据进行处理
    },
  })
);

app.listen(8090, () =>
  console.log('Server listening on http://localhost:8090')
);
